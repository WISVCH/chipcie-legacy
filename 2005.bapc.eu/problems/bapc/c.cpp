/*
  [NKP'05] EVACUATION
  by Jan Kuipers
*/ 

using namespace std;

#include <iostream>
#include <vector>
#include <algorithm>

int Telav, Twait, Tstair;

int calctime (int Tstart, int Felav, int F1, int Fpick, int F2) {
  
  int pickuptime = 0;
  pickuptime >?= (F1-Fpick) * Tstair;
  pickuptime >?= (Fpick-F2) * Tstair;
  pickuptime >?= Tstart + (Felav-Fpick) * Telav;

  return pickuptime + Twait + (Fpick-F2)*Telav; 
}
	  
int main () {

  int runs;
  cin >> runs;
  while (runs--) {

    int F,N;
    cin >>Telav>>Twait>>Tstair;
    cin >>F>>N;

    N+=2;
    vector<int> f(N);
    for (int i=0; i<N-2; i++) cin >>f[i];
    f[N-2]=F;
    f[N-1]=0;
    
    sort(f.begin(),f.end(),greater<int>());

    vector<int> t(N+1,INT_MAX);
    t[0]=0;
		     
    for (int to=1; to<N; to++) {
      int pick=f[to];

      for (int fr=to; fr>=1; fr--) {
	while (calctime(t[fr-1],f[fr-1],f[fr],pick+1,f[to]) <
	       calctime(t[fr-1],f[fr-1],f[fr],pick,f[to])) pick++;

	t[to] <?= calctime(t[fr-1],f[fr-1],f[fr],pick,f[to]);
      }
    }

    int res = f[1]*Tstair;
    for (int n=1; n<N; n++)
      res <?= max(t[n] + f[n]*Telav, f[n+1]*Tstair);
    cout << res << endl;
  }

  return 0;
}
