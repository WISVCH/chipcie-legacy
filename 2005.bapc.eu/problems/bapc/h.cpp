/*
  [NKP'05] Tafeltjes
  door: Jan Kuipers
*/

using namespace std;

#include <iostream>
#include <vector>

const int M = 22;
const int dx[2][3] = { { 1,-1, 1 }, { 1,-1,-1 } };
const int dy[2][3] = { { 0, 0,-1 }, { 0, 0, 1 } };

int N,num,cnt;
int isfree[M][M];

bool isok (int x, int y, int x1, int y1, int x2, int y2) {
  if (y1==y2) return false;
  if (max(y1,y2)<y || min(y1,y2)>y) return false;
  return ((y2-y)*x1 + (y-y1)*x2)/(y2-y1) < x;
}

void go (int dep) {

  if (dep == num/3) {
    cnt++;
    return;
  }

  int best=99, bestx=-1, besty=-1;
  
  for (int x1=0; x1<M; x1++)
    for (int y1=0; y1<M; y1++) {
      if (isfree[x1][y1] == 0) continue;
      
      int tmp=0, type=x1%2, numfree=0;
      
      for (int d1=0; d1<3; d1++) {
	int x2=x1+dx[type][d1], y2=y1+dy[type][d1];
	if (isfree[x2][y2]) {
	  numfree++;
	  for (int d2=0; d2<3; d2++) {
	    int x3=x2+dx[1-type][d2], y3=y2+dy[1-type][d2];
	    if (x3==x1 && y3==y1) continue;
	    if (isfree[x3][y3]) tmp++;
	  }
	}
      }

      if (numfree==2) tmp+=1;
      if (numfree==3) tmp+=3;
      
      if (tmp<best) {
	best=tmp;
	bestx=x1;
	besty=y1;
      }
    }
    
  if (best==99) return;
  
  int x1=bestx,y1=besty,type=bestx%2;
  
  for (int d1=0; d1<3; d1++) {
    int x2=x1+dx[type][d1], y2=y1+dy[type][d1];
    if (isfree[x2][y2]) {
      for (int d2=0; d2<3; d2++) {
	int x3=x2+dx[1-type][d2], y3=y2+dy[1-type][d2];
	if (x3==x1 && y3==y1) continue;

	if (isfree[x3][y3]) {
	  isfree[x1][y1]=isfree[x2][y2]=isfree[x3][y3]=0;
	  go(dep+1);
	  isfree[x1][y1]=isfree[x2][y2]=isfree[x3][y3]=1;
	}
      }
    }
  }

  for (int d1=0; d1<3; d1++)
    for (int d2=d1+1; d2<3; d2++) {
      int x2=x1+dx[type][d1], y2=y1+dy[type][d1];
      int x3=x1+dx[type][d2], y3=y1+dy[type][d2];
      if (isfree[x2][y2] && isfree[x3][y3]) {
	isfree[x1][y1]=isfree[x2][y2]=isfree[x3][y3]=0;
	go(dep+1);
	isfree[x1][y1]=isfree[x2][y2]=isfree[x3][y3]=1;
      }
    }
}

int main () {

  int runs;
  cin >> runs;

  while (runs--) {
    
    cin >> N;
    vector<int> x(N), y(N);
    for (int i=0; i<N; i++) {
      int tmp;
      cin >> tmp;
      x[i] = 3*(tmp/100);
      y[i] = 3*(tmp%100/10);
    }
      
    for (int i=0; i<M; i++)
      for (int j=0; j<M; j++)
	isfree[i][j]=0;

    num=0;
    
    for (int i=0; i<27; i++)
      for (int j=0; j<27; j++) {
	if (i%3==0 || j%3==0 || (i+j)%3==0) continue;

	int sect=0;
	for (int k=0; k<N; k++)
	  if (isok(i,j,x[k],y[k],x[(k+1)%N],y[(k+1)%N])) sect++;

	if (sect%2==1) {
	  isfree[i/3*2+i%3-1+2][j/3+2]=1;
	  num++;
	}
      }
    
    cnt=0;
    if (num%3==0) go(0);
    cout << cnt << endl;
  }
  
  return 0;
}
