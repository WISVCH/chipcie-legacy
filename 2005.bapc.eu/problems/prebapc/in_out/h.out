0
71
104
2676
3763
895402
81267
50541
