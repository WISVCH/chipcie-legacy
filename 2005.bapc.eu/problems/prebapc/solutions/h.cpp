/*
  [NKP'05] Venus Rover
  by: Jan Kuipers
*/

using namespace std;

#include <iostream>
#include <vector>

int main () {

  int runs;
  cin >> runs;
  while (runs--) {
    int N,T,M;
    cin >>N>>T>>M;
    vector<int> t(N),m(N),val(N);
    for (int i=0; i<N; i++) cin>>t[i]>>m[i]>>val[i];

    vector<vector<int> > best(T+1, vector<int>(M+1, 0));
    int res=0;

    for (int i=0; i<N; i++)
      for (int tt=T; tt>=0; tt--)
	for (int mm=M; mm>=0; mm--)
	  if (tt-t[i]>=0 && mm-m[i]>=0)
	    res >?= best[tt][mm] >?= best[tt-t[i]][mm-m[i]] + val[i];

    cout << res << endl;
  }

  return 0;
}
