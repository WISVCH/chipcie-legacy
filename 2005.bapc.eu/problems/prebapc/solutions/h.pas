(*
 * VenusRover
 *
 * Descripton: Example Solution for the VenusRover Task (DAPC/BAPC 2005)
 * Author:     Bram Kuijvenhoven
 *
 *)

program venusrover;

{$mode objfpc}{$H+}

uses
  Classes, SysUtils;

const
  MAX_N = 100;
  MAX_T = 100;
  MAX_M = 100;
  
type
  TStone = record
    t,m,v:integer;
  end;
  
var
  numTestCases,testCase:integer;
  n,t,m:integer;
  stone:array[1..MAX_N] of TStone;
  maxValue:array[0..MAX_T,0..MAX_M] of integer;
  i,j,k:integer;
  max:integer;

begin

  readln(numTestCases);
  for testCase:=1 to numTestCases do
    begin
      // read test case
      readln(n,t,m);
      for i:=1 to n do
        readln(stone[i].t,stone[i].m,stone[i].v);

      // init DP table
      for i:=0 to t do
        for j:=0 to m do
          maxValue[i,j]:=-1;
      maxValue[0,0]:=0;
      
      // fill DP table (2D knapsack algorithm)
      for i:=1 to n do
        begin
          for j:=t downto stone[i].t do
            for k:=m downto stone[i].m do
              if maxValue[j-stone[i].t,k-stone[i].m]>=0 then
                if maxValue[j,k]<maxValue[j-stone[i].t,k-stone[i].m]+stone[i].v then
                  maxValue[j,k]:=maxValue[j-stone[i].t,k-stone[i].m]+stone[i].v;
        end;
        
      // find maximum in DP table
      max:=0;
      for i:=0 to t do
        for j:=0 to m do
          if max<maxValue[i,j] then
            max:=maxValue[i,j];
            
      // output
      writeln(max);
    end;

end.

