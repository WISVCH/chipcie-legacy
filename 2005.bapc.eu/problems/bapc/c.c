#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAXFLOORS 1000
#define true 1
#define false 0
#define INFINITY 1000000000
#define MAX(a, b) ((a) > (b) ? (a) : (b))
#define MIN(a, b) ((a) < (b) ? (a) : (b))

	char isWaiting[MAXFLOORS+1];
	int totalTime[MAXFLOORS+1][MAXFLOORS+2];
	int totalWaitTime[MAXFLOORS+1][MAXFLOORS+2];
	char willNotWait[MAXFLOORS+1][MAXFLOORS+2];

void solveTest() {
	int M, S, W;
	int Nf, Nw;
	int i, f, p;
	int maxfloor = 0;

	scanf("%d %d %d", &M, &S, &W);
	scanf("%d %d", &Nf, &Nw);

	memset(isWaiting, 0, (MAXFLOORS+1)*sizeof(char));
	for (i = 0; i < Nw; i++) {
		int tmp;
		scanf("%d", &tmp);
		maxfloor = MAX(maxfloor, tmp);
		if (tmp)
			isWaiting[tmp] = true;
	}

	if (maxfloor == 0) {
		printf("%d\n", 0);
		return;
	}

	// reset top row
	for (i = 0; i <= Nf; i++) {
		totalTime[i][Nf+1] = (Nf-i) * M;
		totalWaitTime[i][Nf+1] = (Nf-i) * M + S;
		willNotWait[i][Nf+1] = true;
	}

	// fill left colum
	for (i = Nf; i >= 0; i--) {
		if (isWaiting[i]) {
			totalWaitTime[Nf][i] = MAX(totalWaitTime[Nf][i+1], W*(Nf-i) + S);
			willNotWait[Nf][i] = false;
		} else {
			if (willNotWait[Nf][i+1]) {
				totalTime[Nf][i] = totalTime[Nf][i+1];
				totalWaitTime[Nf][i] = totalWaitTime[Nf][i+1];
				willNotWait[Nf][i] = true;
			} else {
				totalWaitTime[Nf][i] = totalWaitTime[Nf][i+1];
				willNotWait[Nf][i] = false;
			}
		}
	}


	for (p = Nf; p >= 0; p--) {
		for (f = Nf-1; f >= 0; f--) {
			int minWait = INFINITY, minNotWait = INFINITY;
			minNotWait = totalWaitTime[f+1][p] + M;
			if (willNotWait[f+1][p]) {
				int t = totalTime[f+1][p] + M;
				if (t < minNotWait) {
					minNotWait = t;
				}
			}
			if (willNotWait[f][p+1] && !isWaiting[p]) {
				if (totalTime[f][p+1] < minNotWait) {
					minNotWait = totalTime[f][p+1];
				}
			}
			willNotWait[f][p] = true;
			totalTime[f][p] = minNotWait;
			if (isWaiting[p]) {
				minWait = MAX(minNotWait, abs(f-p)*W)+S;
 			} else {
				minWait = minNotWait + S;
			}
			if (willNotWait[f][p+1]) {
				int t;
				if (isWaiting[p]) {
					t = MAX(totalTime[f][p+1], abs(f-p)*W)+S;
				} else {
					t = totalTime[f][p+1] + S;
				}
				if (t < minWait) {
					minWait = t; 
				}
			}
			int t = MAX(totalWaitTime[f][p+1], abs(f-p)*W+S);
			if (t < minWait) {
				minWait = t;
			}
			totalWaitTime[f][p] = minWait;
		}
	}

	int ans = maxfloor*W;
	for (f = 0; f <= Nf; f++) {
		if (willNotWait[0][f]) ans = MIN(ans, MAX(totalTime[0][f], ((f-1)*W)));
		ans = MIN(ans, MAX(totalWaitTime[0][f]-S, ((f-1)*W)));
	}
	
	printf("%d\n", ans);
}

int main() {
	int N;
	scanf("%d", &N);
	while (N--) {
		solveTest();
	}
	return 0;
}
