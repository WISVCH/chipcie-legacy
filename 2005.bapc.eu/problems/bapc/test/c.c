/* coneasoup.c */

#include <stdio.h>

int main (void)
{
	FILE *fp;
	if (!(fp = fopen("c.in", "r")))
		exit(1);
	
	int n, i, j;
	char w[255];
	
	fscanf(fp, "%d", &n);
	for (i = 0; i < n; i++)
	{
		fscanf (fp, "%s", &w);
		j = 0;
		while (w[j] != '\0')
		{
			printf("%c", w[j] > 'M' ? w[j] - 13 : w[j] + 13);
			j++;
		}
		printf ("\n");
	}
	return 0;
}
