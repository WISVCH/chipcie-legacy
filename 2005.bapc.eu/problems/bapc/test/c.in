8
BAPC
DELFT
PUVCPVR
CEBOYRZ
FBYHGVBA
VAPERQVOYLYBATJBEQGUNGFUBHYQABGORNYYBJRQ
K
ABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTU