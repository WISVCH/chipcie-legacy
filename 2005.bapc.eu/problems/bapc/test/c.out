ONCP
QRYSG
CHIPCIE
PROBLEM
SOLUTION
INCREDIBLYLONGWORDTHATSHOULDNOTBEALLOWED
X
NOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGHIJKLMNOPQRSTUVWXYZABCDEFGH
